DuckDuckGo/NetworkLeaderboard.swift

```diff
@@ -27,40 +27,65 @@ class NetworkLeaderboard {
 
     struct EntityNames {
 
-        static let visitedSite = "PPVisitedSite"
+        static let pageStats = "PPPageStats"
         static let trackerNetwork = "PPTrackerNetwork"
 
     }
 
-    struct Constants {
-
-        static let startDateKey = "com.duckduckgo.network.leaderboard.start.date"
-
-    }
-
     private lazy var container = DDGPersistenceContainer(name: "NetworkLeaderboard", concurrencyType: .mainQueueConcurrencyType)!
-    private var userDefaults: UserDefaults!
 
     var startDate: Date? {
-        let timeIntervalSince1970 = userDefaults.double(forKey: Constants.startDateKey)
-        guard timeIntervalSince1970 > 0.0 else { return nil }
-        return Date(timeIntervalSince1970: timeIntervalSince1970)
+        return pageStats?.startDate
     }
-
-    init(userDefaults: UserDefaults = UserDefaults()) {
-        self.userDefaults = userDefaults
+    
+    private var pageStats: PPPageStats? {
+        let request: NSFetchRequest<PPPageStats> = PPPageStats.fetchRequest()
+        return try? container.managedObjectContext.fetch(request).first
     }
-
+    
+    init() {
+        if pageStats == nil {
+            reset()
+        }
+    }
+    
     func reset() {
-        container.deleteAll(entities: try? container.managedObjectContext.fetch(PPVisitedSite.fetchRequest()))
         container.deleteAll(entities: try? container.managedObjectContext.fetch(PPTrackerNetwork.fetchRequest()))
+        container.deleteAll(entities: try? container.managedObjectContext.fetch(PPPageStats.fetchRequest()))
+        createNewPageStatsEntity()
         _ = container.save()
-        userDefaults.removeObject(forKey: Constants.startDateKey)
     }
 
-    func sitesVisited() -> Int {
-        let request: NSFetchRequest<PPVisitedSite> = PPVisitedSite.fetchRequest()
-        return (try? container.managedObjectContext.count(for: request)) ?? 0
+    func pageVisited() {
+        if let pageStats = pageStats {
+            let count = (pageStats.pagesLoaded ?? 0).intValue
+            pageStats.pagesLoaded = NSNumber(value: count + 1)
+            _ = container.save()
+        }
+    }
+    
+    func pageHasTrackers() {
+        if let pageStats = pageStats {
+            let count = (pageStats.pagesWithTrackers ?? 0).intValue
+            pageStats.pagesWithTrackers = NSNumber(value: count + 1)
+            _ = container.save()
+        }
+    }
+    
+    private func createNewPageStatsEntity() {
+        let managedObject = NSEntityDescription.insertNewObject(forEntityName: EntityNames.pageStats, into: container.managedObjectContext)
+        guard let stats = managedObject as? PPPageStats else { return }
+        stats.startDate = Date()
+        stats.pagesLoaded = NSNumber(value: 0)
+        _ = container.save()
+    }
+    
+    func pagesVisited() -> Int {
+        return pageStats?.pagesLoaded?.intValue ?? 0
+    }
+    
+    func pagesWithTrackers() -> Int {
+        return pageStats?.pagesWithTrackers?.intValue ?? 0
```

Comment:
There is a bit of clash in my mind when looking at 'pageVisited' and 'pagesVisited' function invocations.

Maybe we can rename 'pagesVisited' -> 'visitedPages(Count)' and 'pagesWithTrackers' -> 'visitedPagesWithTrackers(Count)' to make it a bit more explicit?